File: backend/src/respira_ally/core/schemas/task.py
"""
Task Schemas
Pydantic models for Task Management System API endpoints

Sprint 5: Task Management System - Therapist action items and assignments
"""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ============================================================================
# Task Enums (mirrored from Domain Entity)
# ============================================================================


class TaskPriority(str, Enum):
    """Task Priority Levels - Based on Alert Severity and GOLD ABE classification"""

    CRITICAL = "CRITICAL"  # Alert: CRITICAL severity
    HIGH = "HIGH"  # Alert: HIGH severity or GOLD Group E
    MEDIUM = "MEDIUM"  # Alert: MEDIUM severity
    LOW = "LOW"  # Alert: LOW severity or routine tasks


class TaskStatus(str, Enum):
    """Task Status Lifecycle"""

    TODO = "TODO"  # Task created, not started
    IN_PROGRESS = "IN_PROGRESS"  # Task assigned and in progress
    DONE = "DONE"  # Task completed
    CANCELLED = "CANCELLED"  # Task cancelled


class TaskType(str, Enum):
    """Task Type - Auto-generated or Manual"""

    ALERT_TRIGGERED = "ALERT_TRIGGERED"  # Auto-generated from Alert
    MANUAL = "MANUAL"  # Manually created by therapist
    SCHEDULED = "SCHEDULED"  # Scheduled routine task


# ============================================================================
# Task Request Schemas
# ============================================================================


class TaskCreateRequest(BaseModel):
    """
    Request schema for creating a new task (POST /api/v1/tasks/)

    Therapists can manually create tasks for patients.
    """

    model_config = ConfigDict(from_attributes=True)

    # Required fields
    patient_id: UUID = Field(..., description="Patient UUID")
    title: str = Field(..., min_length=1, max_length=200, description="Task title")
    priority: TaskPriority = Field(..., description="Task priority level")
    task_type: TaskType = Field(
        default=TaskType.MANUAL, description="Task type (defaults to MANUAL)"
    )

    # Optional fields
    description: Optional[str] = Field(None, description="Detailed task description")
    assigned_to: Optional[UUID] = Field(None, description="Therapist UUID to assign to")
    related_alert_id: Optional[UUID] = Field(
        None, description="Related alert ID (if triggered by alert)"
    )
    due_date: Optional[datetime] = Field(None, description="Task due date (ISO 8601)")
    task_metadata: Optional[dict] = Field(
        None, description="Additional task context (GOLD group, scores, etc.)"
    )


class TaskUpdateRequest(BaseModel):
    """
    Request schema for updating an existing task (PATCH /api/v1/tasks/{id})

    Therapists can update task details, assignment, and metadata.
    Status transitions should use dedicated endpoints (/start, /complete, /cancel).
    """

    model_config = ConfigDict(from_attributes=True)

    # All fields optional for partial updates
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None)
    priority: Optional[TaskPriority] = Field(None)
    assigned_to: Optional[UUID] = Field(None, description="Reassign to different therapist")
    due_date: Optional[datetime] = Field(None)
    task_metadata: Optional[dict] = Field(None)


class TaskAssignRequest(BaseModel):
    """
    Request schema for assigning a task to a therapist (POST /api/v1/tasks/{id}/assign)
    """

    model_config = ConfigDict(from_attributes=True)

    therapist_id: UUID = Field(..., description="Therapist UUID to assign to")


class TaskCancelRequest(BaseModel):
    """
    Request schema for cancelling a task (POST /api/v1/tasks/{id}/cancel)
    """

    model_config = ConfigDict(from_attributes=True)

    reason: Optional[str] = Field(None, description="Cancellation reason")


# ============================================================================
# Task Response Schemas
# ============================================================================


class TaskResponse(BaseModel):
    """
    Task response schema for API endpoints

    Represents a single task with all details.
    """

    model_config = ConfigDict(from_attributes=True)

    # Core fields
    task_id: UUID = Field(..., description="Task UUID")
    patient_id: UUID = Field(..., description="Patient UUID")
    title: str = Field(..., description="Task title")
    description: Optional[str] = Field(None, description="Task description")

    # Classification
    task_type: TaskType = Field(..., description="Task type")
    priority: TaskPriority = Field(..., description="Task priority level")
    status: TaskStatus = Field(..., description="Task status")

    # Assignment and relationships
    assigned_to: Optional[UUID] = Field(None, description="Assigned therapist UUID")
    related_alert_id: Optional[UUID] = Field(None, description="Related alert UUID")

    # Metadata
    task_metadata: Optional[dict] = Field(None, description="Additional task context")

    # Timestamps
    due_date: Optional[str] = Field(None, description="Due date (ISO 8601)")
    completed_at: Optional[str] = Field(None, description="Completion timestamp (ISO 8601)")
    created_at: str = Field(..., description="Creation timestamp (ISO 8601)")
    updated_at: str = Field(..., description="Last update timestamp (ISO 8601)")

    # Computed fields
    is_overdue: bool = Field(..., description="Whether task is overdue")


class TaskListResponse(BaseModel):
    """
    Paginated task list response

    Used for GET /api/v1/patients/{patient_id}/tasks endpoint
    """

    model_config = ConfigDict(from_attributes=True)

    tasks: list[TaskResponse] = Field(..., description="List of tasks")
    total: int = Field(..., description="Total number of tasks matching filters")
    page: int = Field(..., description="Current page number (0-indexed)")
    page_size: int = Field(..., description="Number of items per page")
    total_pages: int = Field(..., description="Total number of pages")


# ============================================================================
# Task Statistics Schemas
# ============================================================================


class TaskStatsResponse(BaseModel):
    """
    Task statistics for a patient or therapist

    Provides overview of task distribution and completion metrics.
    """

    model_config = ConfigDict(from_attributes=True)

    # Scope identifiers
    patient_id: Optional[UUID] = Field(None, description="Patient UUID (if patient stats)")
    therapist_id: Optional[UUID] = Field(None, description="Therapist UUID (if therapist stats)")

    # Status breakdown
    total_tasks: int = Field(..., description="Total tasks")
    todo_count: int = Field(..., description="TODO tasks")
    in_progress_count: int = Field(..., description="IN_PROGRESS tasks")
    done_count: int = Field(..., description="DONE tasks")
    cancelled_count: int = Field(..., description="CANCELLED tasks")

    # Priority breakdown
    critical_count: int = Field(0, description="CRITICAL priority tasks")
    high_count: int = Field(0, description="HIGH priority tasks")
    medium_count: int = Field(0, description="MEDIUM priority tasks")
    low_count: int = Field(0, description="LOW priority tasks")

    # Overdue tasks
    overdue_count: int = Field(0, description="Overdue tasks")

    # Last activity
    last_task_date: Optional[datetime] = Field(None, description="Last task created date")
    last_task_title: Optional[str] = Field(None, description="Last task title")

File: frontend/dashboard/lib/types/task.ts
/**
 * Task Type Definitions
 * TypeScript interfaces mirroring backend Pydantic schemas
 *
 * Backend Reference:
 * - backend/src/respira_ally/core/schemas/task.py
 * - backend/src/respira_ally/domain/entities/task.py
 *
 * Sprint 5 - Task Board UI
 */

// ============================================================================
// Task Enums (mirrored from backend)
// ============================================================================

/**
 * Task Status Lifecycle
 * Represents the current state of a task
 */
export enum TaskStatus {
  TODO = 'TODO',                // Task created, not started
  IN_PROGRESS = 'IN_PROGRESS',  // Task assigned and in progress
  DONE = 'DONE',                // Task completed
  CANCELLED = 'CANCELLED',      // Task cancelled (hidden from board)
}

/**
 * Task Priority Levels
 * Based on Alert Severity and GOLD ABE classification
 */
export enum TaskPriority {
  CRITICAL = 'CRITICAL',  // Alert: CRITICAL severity
  HIGH = 'HIGH',          // Alert: HIGH severity or GOLD Group E
  MEDIUM = 'MEDIUM',      // Alert: MEDIUM severity
  LOW = 'LOW',            // Alert: LOW severity or routine tasks
}

/**
 * Task Type - Auto-generated or Manual
 */
export enum TaskType {
  ALERT_TRIGGERED = 'ALERT_TRIGGERED',  // Auto-generated from Alert
  MANUAL = 'MANUAL',                    // Manually created by therapist
  SCHEDULED = 'SCHEDULED',              // Scheduled routine task
}

// ============================================================================
// Task Interfaces
// ============================================================================

/**
 * Task Response Schema
 * Represents a single task with all details
 *
 * Maps to backend TaskResponse schema
 */
export interface Task {
  // Core fields
  task_id: string           // UUID
  patient_id: string        // Patient UUID
  title: string             // Task title (1-200 chars)
  description?: string      // Task description (optional)

  // Classification
  task_type: TaskType       // Task type
  priority: TaskPriority    // Task priority level
  status: TaskStatus        // Task status

  // Assignment and relationships
  assigned_to?: string      // Assigned therapist UUID (optional)
  related_alert_id?: string // Related alert UUID (optional)

  // Metadata
  task_metadata?: Record<string, any> // Additional task context (GOLD group, scores, etc.)

  // Timestamps (ISO 8601 format)
  due_date?: string         // Due date (optional)
  completed_at?: string     // Completion timestamp (optional)
  created_at: string        // Creation timestamp
  updated_at: string        // Last update timestamp

  // Computed fields
  is_overdue: boolean       // Whether task is overdue
}

/**
 * Task List Response (Paginated)
 * Used for GET /api/v1/patients/{patient_id}/tasks endpoint
 */
export interface TaskListResponse {
  tasks: Task[]             // List of tasks
  total: number             // Total number of tasks matching filters
  page: number              // Current page number (0-indexed)
  page_size: number         // Number of items per page
  total_pages: number       // Total number of pages
}

/**
 * Task Statistics Response
 * Provides overview of task distribution and completion metrics
 */
export interface TaskStatsResponse {
  // Scope identifiers
  patient_id?: string       // Patient UUID (if patient stats)
  therapist_id?: string     // Therapist UUID (if therapist stats)

  // Status breakdown
  total_tasks: number       // Total tasks
  todo_count: number        // TODO tasks
  in_progress_count: number // IN_PROGRESS tasks
  done_count: number        // DONE tasks
  cancelled_count: number   // CANCELLED tasks

  // Priority breakdown
  critical_count: number    // CRITICAL priority tasks
  high_count: number        // HIGH priority tasks
  medium_count: number      // MEDIUM priority tasks
  low_count: number         // LOW priority tasks

  // Overdue tasks
  overdue_count: number     // Overdue tasks

  // Last activity
  last_task_date?: string   // Last task created date (ISO 8601)
  last_task_title?: string  // Last task title
}

// ============================================================================
// Task Request Schemas (for API calls)
// ============================================================================

/**
 * Task Create Request
 * Used for POST /api/v1/tasks endpoint
 */
export interface TaskCreateRequest {
  // Required fields
  patient_id: string
  title: string
  priority: TaskPriority
  task_type?: TaskType      // Defaults to MANUAL

  // Optional fields
  description?: string
  assigned_to?: string      // Therapist UUID
  related_alert_id?: string // Alert UUID
  due_date?: string         // ISO 8601
  task_metadata?: Record<string, any>
}

/**
 * Task Update Request
 * Used for PATCH /api/v1/tasks/{id} endpoint
 * All fields optional for partial updates
 */
export interface TaskUpdateRequest {
  title?: string
  description?: string
  priority?: TaskPriority
  assigned_to?: string      // Reassign to different therapist
  due_date?: string
  task_metadata?: Record<string, any>
}

/**
 * Task Assign Request
 * Used for POST /api/v1/tasks/{id}/assign endpoint
 */
export interface TaskAssignRequest {
  therapist_id: string      // Therapist UUID to assign to
}

/**
 * Task Cancel Request
 * Used for POST /api/v1/tasks/{id}/cancel endpoint
 */
export interface TaskCancelRequest {
  reason?: string           // Cancellation reason (optional)
}

// ============================================================================
// Task Filters (for UI filtering)
// ============================================================================

/**
 * Task Filters
 * Used for client-side filtering in TaskBoard
 */
export interface TaskFilters {
  priority?: TaskPriority   // Filter by priority
  task_type?: TaskType      // Filter by task type
  show_overdue_only?: boolean // Show only overdue tasks
}

// ============================================================================
// Priority Color Configurations
// ============================================================================

/**
 * Priority Color Config
 * Color scheme for priority badges and cards
 */
export interface PriorityColorConfig {
  bg: string        // Background color (Tailwind class)
  border: string    // Border color (Tailwind class)
  text: string      // Text color (Tailwind class)
  badge: string     // Badge color (Tailwind class)
}

/**
 * Priority color mapping
 * Used for visual representation of task priorities
 */
export const PRIORITY_COLORS: Record<TaskPriority, PriorityColorConfig> = {
  [TaskPriority.CRITICAL]: {
    bg: 'bg-red-50',
    border: 'border-red-500',
    text: 'text-red-700',
    badge: 'bg-red-500 text-white',
  },
  [TaskPriority.HIGH]: {
    bg: 'bg-orange-50',
    border: 'border-orange-500',
    text: 'text-orange-700',
    badge: 'bg-orange-500 text-white',
  },
  [TaskPriority.MEDIUM]: {
    bg: 'bg-yellow-50',
    border: 'border-yellow-500',
    text: 'text-yellow-700',
    badge: 'bg-yellow-500 text-white',
  },
  [TaskPriority.LOW]: {
    bg: 'bg-blue-50',
    border: 'border-blue-500',
    text: 'text-blue-700',
    badge: 'bg-blue-500 text-white',
  },
}

// ============================================================================
// Task Type Icons
// ============================================================================

/**
 * Task type icon mapping
 */
export const TASK_TYPE_ICONS: Record<TaskType, string> = {
  [TaskType.ALERT_TRIGGERED]: '🔔',  // Auto-generated from alert
  [TaskType.MANUAL]: '✏️',            // Manually created
  [TaskType.SCHEDULED]: '📅',         // Scheduled routine
}

// ============================================================================
// Status Display Names (Traditional Chinese)
// ============================================================================

/**
 * Status display names in Traditional Chinese
 */
export const TASK_STATUS_LABELS: Record<TaskStatus, string> = {
  [TaskStatus.TODO]: '待處理',
  [TaskStatus.IN_PROGRESS]: '進行中',
  [TaskStatus.DONE]: '已完成',
  [TaskStatus.CANCELLED]: '已取消',
}

/**
 * Priority display names in Traditional Chinese
 */
export const TASK_PRIORITY_LABELS: Record<TaskPriority, string> = {
  [TaskPriority.CRITICAL]: '緊急',
  [TaskPriority.HIGH]: '高',
  [TaskPriority.MEDIUM]: '中',
  [TaskPriority.LOW]: '低',
}

/**
 * Task type display names in Traditional Chinese
 */
export const TASK_TYPE_LABELS: Record<TaskType, string> = {
  [TaskType.ALERT_TRIGGERED]: '警示觸發',
  [TaskType.MANUAL]: '手動建立',
  [TaskType.SCHEDULED]: '排程任務',
}

// ============================================================================
// Helper Type Guards
// ============================================================================

/**
 * Check if a task is overdue
 */
export function isTaskOverdue(task: Task): boolean {
  if (!task.due_date) return false
  const dueDate = new Date(task.due_date)
  const now = new Date()
  return dueDate < now && task.status !== TaskStatus.DONE
}

/**
 * Check if a task can be started
 */
export function canStartTask(task: Task): boolean {
  return task.status === TaskStatus.TODO
}

/**
 * Check if a task can be completed
 */
export function canCompleteTask(task: Task): boolean {
  return task.status === TaskStatus.IN_PROGRESS
}

/**
 * Check if a task can be cancelled
 */
export function canCancelTask(task: Task): boolean {
  return task.status === TaskStatus.TODO || task.status === TaskStatus.IN_PROGRESS
}

/**
 * Get days until due date
 * Returns negative number if overdue
 */
export function getDaysUntilDue(task: Task): number | null {
  if (!task.due_date) return null
  const dueDate = new Date(task.due_date)
  const now = new Date()
  const diffTime = dueDate.getTime() - now.getTime()
  const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24))
  return diffDays
}

/**
 * Format due date display
 * Examples: "今天到期", "明天到期", "2 天後到期", "逾期 3 天"
 */
export function formatDueDate(task: Task): string {
  if (!task.due_date) return ''

  const days = getDaysUntilDue(task)
  if (days === null) return ''

  if (days === 0) return '今天到期'
  if (days === 1) return '明天到期'
  if (days > 1) return `${days} 天後到期`
  if (days === -1) return '逾期 1 天'
  return `逾期 ${Math.abs(days)} 天`
}
